pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -796,91 +793,58 @@ func createHostDnsmasqFile(ctx *zedrouterContext, bridge string) {
-func lookupOrAllocateIPv4(
-	ctx *zedrouterContext,
-	status *types.NetworkInstanceStatus,
-	mac net.HardwareAddr) (string, error) {
+// Returns an IP address as a string, or "" on error
+func lookupOrAllocateIPv4(status *types.NetworkInstanceStatus,
+	appID uuid.UUID, appNum int, mac net.HardwareAddr) (string, error) {
+	log.Functionf("lookupOrAllocateIPv4(%s-%s): appNum:%d\n",
+		status.DisplayName, status.Key(), appNum)
 
-	log.Functionf("lookupOrAllocateIPv4(%s-%s): mac:%s\n",
-		status.DisplayName, status.Key(), mac.String())
-	// Lookup to see if it exists
-	if ip, ok := status.IPAssignments[mac.String()]; ok {
-		log.Functionf("found Ip addr ( %s) for mac(%s)\n",
-			ip.String(), mac.String())
+	networkID := status.UUIDandVersion.UUID
+	// if it exists
+	if ip := getIPAddrAssignment(status, mac); ip != nil {
+		log.Functionf("lookupOrAllocateIPv4(%s-%s): found %s for %s\n",
+			networkID.String(), appID.String(), ip.String(), mac.String())
 		return ip.String(), nil
 	}
-
-	log.Functionf("bridgeName %s Subnet %v range %v-%v\n",
-		status.BridgeName, status.Subnet,
-		status.DhcpRange.Start.String(), status.DhcpRange.End.String())
-
 	if status.DhcpRange.Start == nil {
 		if status.Type == types.NetworkInstanceTypeSwitch {
-			log.Functionf("%s-%s switch means no bridgeIpAddr",
+			log.Functionf("%s-%s switch means no IPAddr",
 				status.DisplayName, status.Key())
 			return "", nil
 		}
 		log.Fatalf("%s-%s: nil DhcpRange.Start",
 			status.DisplayName, status.Key())
 	}
 
-	// Starting guess based on number allocated
-	allocated := uint(len(status.IPAssignments))
-	if status.Gateway != nil {
-		// With Gateway present in network instance status,
-		// we would have used that as our Bridge IP address and not
-		// allocated new one. Since bridge IP address is also stored
-		// as part of IPAssignments, the actual allocated IP address
-		// numner is 1 less than the length of IPAssignments map size.
-		allocated--
-	}
-	a := addToIP(status.DhcpRange.Start, allocated)
-	for status.DhcpRange.End == nil ||
-		bytes.Compare(a, status.DhcpRange.End) <= 0 {
-
-		log.Functionf("lookupOrAllocateIPv4(%s) testing %s\n",
-			mac.String(), a.String())
-		if status.IsIpAssigned(a) {
-			a = addToIP(a, 1)
-			continue
-		}
-		log.Functionf("lookupOrAllocateIPv4(%s) found free %s\n",
-			mac.String(), a.String())
+	// get ip address
+	a := types.AddToIP(status.DhcpRange.Start, appNum)
 
-		recordIPAssignment(ctx, status, a, mac.String())
-		return a.String(), nil
+	ret, err := status.DhcpRange.Contains(a)
+	if err != nil {
+		log.Errorf("lookupOrAllocateIPv4(%s, %s): fail: %s\n",
+			networkID.String(), appID.String(), err)
+		return "", err
+	}
+	if !ret {
+		errStr := fmt.Sprintf("AppIPAddr(%s) is not in DhcpRange(%s, %s)",
+			a.String(), status.DhcpRange.Start.String(),
+			status.DhcpRange.End.String())
+		log.Errorf("lookupOrAllocateIPv4(%s, %s): fail: %s",
+			networkID.String(), appID.String(), errStr)
```

Comment:
This is not a friendly error message; this happens when there are no more IP addresses available in the dhcp range, hence the error message should say that. User doesn't care which functions or variables are used.
Thus errStr := "no free IP addresses in DHCP range" is a better error message.